File: go/fixture/oauth2_token_provider.go
// Copyright (c) 2026 Peaceful Studio OÜ
// SPDX-License-Identifier: Apache-2.0

package fixture

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// RefreshSkew is the safety margin subtracted from the access token's nominal
// expiry. A token is refreshed once its remaining lifetime drops below this
// threshold, so callers never present an about-to-expire token to the API.
const RefreshSkew = 30 * time.Second

// OAuth2TokenProviderConfig configures an OAuth2TokenProvider.
type OAuth2TokenProviderConfig struct {
	TokenURL     string
	ClientID     string
	ClientSecret string
	Audience     string
	Scope        string
	HTTPClient   *http.Client
	Now          func() time.Time
}

// OAuth2TokenProvider obtains and caches OAuth2 access tokens via the
// client_credentials grant. It is safe for concurrent use; in-flight refreshes
// are coalesced so concurrent callers don't trigger redundant token requests.
type OAuth2TokenProvider struct {
	cfg OAuth2TokenProviderConfig

	mu          sync.Mutex
	cached      string
	expiresAt   time.Time
	refreshing  bool
	refreshDone chan struct{}
	refreshErr  error
}

// NewOAuth2TokenProvider constructs an OAuth2TokenProvider.
//
// TokenURL, ClientID and ClientSecret are required. If HTTPClient is nil,
// http.DefaultClient is used. If Now is nil, time.Now is used.
func NewOAuth2TokenProvider(cfg OAuth2TokenProviderConfig) (*OAuth2TokenProvider, error) {
	if strings.TrimSpace(cfg.TokenURL) == "" {
		return nil, errors.New("oauth2: TokenURL is required")
	}
	if strings.TrimSpace(cfg.ClientID) == "" {
		return nil, errors.New("oauth2: ClientID is required")
	}
	if strings.TrimSpace(cfg.ClientSecret) == "" {
		return nil, errors.New("oauth2: ClientSecret is required")
	}
	if cfg.HTTPClient == nil {
		cfg.HTTPClient = http.DefaultClient
	}
	if cfg.Now == nil {
		cfg.Now = time.Now
	}
	return &OAuth2TokenProvider{cfg: cfg}, nil
}

// Token returns a cached access token, refreshing it via the token endpoint
// if the current one is missing or within RefreshSkew of expiry.
func (p *OAuth2TokenProvider) Token(ctx context.Context) (string, error) {
	for {
		p.mu.Lock()
		if p.tokenIsFreshLocked() {
			token := p.cached
			p.mu.Unlock()
			return token, nil
		}
		if p.refreshing {
			done := p.refreshDone
			p.mu.Unlock()
			select {
			case <-done:
			case <-ctx.Done():
				return "", ctx.Err()
			}
			p.mu.Lock()
			leaderErr := p.refreshErr
			cached := p.cached
			p.mu.Unlock()

			if leaderErr == nil {
				return cached, nil
			}
			if isContextError(leaderErr) && ctx.Err() == nil {
				continue
			}
			return "", leaderErr
		}
		p.refreshing = true
		p.refreshDone = make(chan struct{})
		p.refreshErr = nil
		p.mu.Unlock()

		token, expiresAt, err := p.fetchToken(ctx)

		p.mu.Lock()
		if err != nil {
			p.refreshErr = err
		} else {
			p.cached = token
			p.expiresAt = expiresAt
		}
		close(p.refreshDone)
		p.refreshing = false
		p.mu.Unlock()
		if err != nil {
			return "", err
		}
		return token, nil
	}
}

func isContextError(err error) bool {
	return errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded)
}

const errorBodyMaxBytes = 512

func truncateBodyForError(body []byte) string {
	trimmed := strings.TrimSpace(string(body))
	if len(trimmed) <= errorBodyMaxBytes {
		return trimmed
	}
	return trimmed[:errorBodyMaxBytes] + "...[truncated]"
}

// Invalidate drops the cached token. The next call to Token forces a refresh.
func (p *OAuth2TokenProvider) Invalidate() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.cached = ""
	p.expiresAt = time.Time{}
}

func (p *OAuth2TokenProvider) tokenIsFreshLocked() bool {
	if p.cached == "" {
		return false
	}
	return p.cfg.Now().Add(RefreshSkew).Before(p.expiresAt)
}

type tokenResponse struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	ExpiresIn   int    `json:"expires_in"`
}

type tokenErrorResponse struct {
	Error            string `json:"error"`
	ErrorDescription string `json:"error_description"`
}

func (p *OAuth2TokenProvider) fetchToken(ctx context.Context) (string, time.Time, error) {
	form := url.Values{}
	form.Set("grant_type", "client_credentials")
	form.Set("client_id", p.cfg.ClientID)
	form.Set("client_secret", p.cfg.ClientSecret)
	if p.cfg.Audience != "" {
		form.Set("audience", p.cfg.Audience)
	}
	if p.cfg.Scope != "" {
		form.Set("scope", p.cfg.Scope)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.cfg.TokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", time.Time{}, fmt.Errorf("oauth2: build token request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	resp, err := p.cfg.HTTPClient.Do(req)
	if err != nil {
		return "", time.Time{}, fmt.Errorf("oauth2: token request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", time.Time{}, fmt.Errorf("oauth2: read token response: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var er tokenErrorResponse
		_ = json.Unmarshal(body, &er)
		if er.Error != "" {
			if er.ErrorDescription != "" {
				return "", time.Time{}, fmt.Errorf("oauth2: token endpoint returned HTTP %d: %s: %s", resp.StatusCode, er.Error, er.ErrorDescription)
			}
			return "", time.Time{}, fmt.Errorf("oauth2: token endpoint returned HTTP %d: %s", resp.StatusCode, er.Error)
		}
		return "", time.Time{}, fmt.Errorf("oauth2: token endpoint returned HTTP %d: %s", resp.StatusCode, truncateBodyForError(body))
	}

	var tr tokenResponse
	if err := json.Unmarshal(body, &tr); err != nil {
		return "", time.Time{}, fmt.Errorf("oauth2: decode token response: %w", err)
	}
	if tr.AccessToken == "" {
		return "", time.Time{}, errors.New("oauth2: token endpoint returned empty access_token")
	}
	if tr.ExpiresIn <= 0 {
		return "", time.Time{}, fmt.Errorf("oauth2: token endpoint returned non-positive expires_in (%d)", tr.ExpiresIn)
	}
	return tr.AccessToken, p.cfg.Now().Add(time.Duration(tr.ExpiresIn) * time.Second), nil
}

File: csharp/src/Peaceful.Canton.Localnet.Testing/OAuth2TokenProvider.cs
// Copyright (c) 2026 Peaceful Studio OÜ
// SPDX-License-Identifier: Apache-2.0

using System.Net.Http.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Peaceful.Canton.Localnet.Testing;

/// <summary>
/// Settings that drive the <c>client_credentials</c> grant: the token endpoint,
/// the audience to request, and the client id/secret pair. <paramref name="Scope"/>
/// is optional; if omitted, no <c>scope</c> form field is sent.
/// </summary>
public sealed record OAuth2TokenProviderOptions(
    Uri TokenEndpoint,
    string ClientId,
    string ClientSecret,
    string Audience,
    string? Scope = "openid")
{
    /// <summary>
    /// Safety margin used to consider an access token expired before its nominal
    /// expiry. Defaults to 30 seconds so callers don't race a clock-skewed
    /// validator into a 401.
    /// </summary>
    public TimeSpan ExpiryLeeway { get; init; } = TimeSpan.FromSeconds(30);
}

/// <summary>
/// Caches an OAuth2 access token in memory and refreshes it on expiry. A single
/// instance is safe to share across threads; concurrent callers will await one
/// in-flight refresh rather than fan out duplicate token requests.
/// </summary>
public sealed class OAuth2TokenProvider : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly OAuth2TokenProviderOptions _options;
    private readonly ILogger<OAuth2TokenProvider> _logger;
    private readonly TimeProvider _timeProvider;
    private readonly SemaphoreSlim _refreshLock = new(1, 1);

    private CachedToken? _cached;

    /// <summary>
    /// Creates a token provider for a single <c>client_credentials</c> client.
    /// </summary>
    /// <param name="httpClient">Client used to POST to the token endpoint.</param>
    /// <param name="options">Token endpoint, audience, scope, and client credentials.</param>
    /// <param name="logger">Optional logger; defaults to a no-op logger.</param>
    /// <param name="timeProvider">
    /// Clock used for cache-expiry decisions; defaults to
    /// <see cref="TimeProvider.System"/>. Override it in tests to drive expiry
    /// deterministically.
    /// </param>
    public OAuth2TokenProvider(
        HttpClient httpClient,
        OAuth2TokenProviderOptions options,
        ILogger<OAuth2TokenProvider>? logger = null,
        TimeProvider? timeProvider = null)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _options = options ?? throw new ArgumentNullException(nameof(options));
        _logger = logger ?? NullLogger<OAuth2TokenProvider>.Instance;
        _timeProvider = timeProvider ?? TimeProvider.System;
    }

    /// <summary>
    /// Returns a valid access token, refreshing if the cached token is missing
    /// or within the configured leeway of its expiry. Thread-safe.
    /// </summary>
    public async ValueTask<string> GetAccessTokenAsync(CancellationToken cancellationToken = default)
    {
        var cached = _cached;
        var now = _timeProvider.GetUtcNow();
        if (cached is not null && cached.IsValidAt(now, _options.ExpiryLeeway))
        {
            return cached.AccessToken;
        }

        await _refreshLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            cached = _cached;
            now = _timeProvider.GetUtcNow();
            if (cached is not null && cached.IsValidAt(now, _options.ExpiryLeeway))
            {
                return cached.AccessToken;
            }

            var fresh = await FetchAsync(cancellationToken).ConfigureAwait(false);
            _cached = fresh;
            return fresh.AccessToken;
        }
        finally
        {
            _refreshLock.Release();
        }
    }

    private async Task<CachedToken> FetchAsync(CancellationToken cancellationToken)
    {
        var form = new List<KeyValuePair<string, string>>
        {
            new("grant_type", "client_credentials"),
            new("client_id", _options.ClientId),
            new("client_secret", _options.ClientSecret),
            new("audience", _options.Audience),
        };
        if (!string.IsNullOrEmpty(_options.Scope))
        {
            form.Add(new KeyValuePair<string, string>("scope", _options.Scope));
        }

        using var request = new HttpRequestMessage(HttpMethod.Post, _options.TokenEndpoint)
        {
            Content = new FormUrlEncodedContent(form),
        };

        _logger.LogDebug(
            "Requesting OAuth2 token from {TokenEndpoint} for client {ClientId} audience {Audience}",
            _options.TokenEndpoint,
            _options.ClientId,
            _options.Audience);

        using var response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            throw new OAuth2TokenException(
                $"Token endpoint {_options.TokenEndpoint} returned {(int)response.StatusCode} {response.ReasonPhrase}: {body}",
                response.StatusCode,
                body);
        }

        var payload = await response.Content
            .ReadFromJsonAsync<TokenResponse>(cancellationToken: cancellationToken)
            .ConfigureAwait(false)
            ?? throw new OAuth2TokenException(
                $"Token endpoint {_options.TokenEndpoint} returned an empty body.",
                response.StatusCode,
                string.Empty);

        if (string.IsNullOrEmpty(payload.AccessToken))
        {
            throw new OAuth2TokenException(
                $"Token endpoint {_options.TokenEndpoint} returned a response with no access_token.",
                response.StatusCode,
                string.Empty);
        }

        var lifetime = payload.ExpiresIn > 0 ? TimeSpan.FromSeconds(payload.ExpiresIn) : TimeSpan.FromMinutes(5);
        var expiresAt = _timeProvider.GetUtcNow() + lifetime;
        return new CachedToken(payload.AccessToken, expiresAt);
    }

    /// <summary>Releases the internal refresh lock.</summary>
    public void Dispose()
    {
        _refreshLock.Dispose();
    }

    private sealed record CachedToken(string AccessToken, DateTimeOffset ExpiresAtUtc)
    {
        public bool IsValidAt(DateTimeOffset now, TimeSpan leeway) => now + leeway < ExpiresAtUtc;
    }

    private sealed record TokenResponse(
        [property: JsonPropertyName("access_token")] string AccessToken,
        [property: JsonPropertyName("token_type")] string? TokenType,
        [property: JsonPropertyName("expires_in")] int ExpiresIn,
        [property: JsonPropertyName("scope")] string? Scope);
}

/// <summary>
/// Raised when the OAuth2 token endpoint returns a non-success status, an empty
/// body, or a body with no <c>access_token</c>.
/// </summary>
public sealed class OAuth2TokenException : Exception
{
    /// <summary>
    /// Creates the exception with the originating <paramref name="statusCode"/>
    /// and raw <paramref name="responseBody"/> for diagnostics.
    /// </summary>
    /// <param name="message">Human-readable description of the failure.</param>
    /// <param name="statusCode">HTTP status the token endpoint returned.</param>
    /// <param name="responseBody">Raw response body, or empty when none was read.</param>
    public OAuth2TokenException(string message, System.Net.HttpStatusCode statusCode, string responseBody)
        : base(message)
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
    }

    /// <summary>HTTP status the token endpoint returned.</summary>
    public System.Net.HttpStatusCode StatusCode { get; }

    /// <summary>Raw response body, or empty when none was read.</summary>
    public string ResponseBody { get; }
}
